cypress/integration/switch.spec.js

```diff
@@ -0,0 +1,43 @@
+describe('switch v1', () => {
+  it('works', () => {
+    cy.visit('/#/doodad/Switch')
+    cy.percySnapshot()
+    cy.injectAxe()
+    assertInteractive()
+    assertAlwaysOn()
+    assertAlwaysOff()
+  })
+})
+
+function assertInteractive () {
+  cy.checkA11y('[aria-controls=switch-interactive]')
+  cy.get('[aria-controls=switch-interactive]')
+    .should('have.attr', 'aria-checked', 'true')
```

Comment:
My idea is to essentially replace the elm tests entirely with cypress tests. The way I see it, this repository is all front-end, so why not use UI testing? It covers everything you would cover in the elm tests, and then some. I assume that devs would be manually clicking through a widget that they modified anyways to make sure it works. Might as well automate it. Plus, you can bundle all of that with the axe checks and Percy screenshots.
 
About speed: For me, this test took 7 seconds to run. I think that's fast enough to support a development cycle. That is to say, *change some elm code*, *run the cypress test*, *I see a regression after 7 seconds*, *fix it*, *run the cypress test*, *I see it pass in 7 seconds*.

About ergonomics: That's true, although I think the small learning curve is worth it. And if all they'd be doing is maintaining, then it's nearly effortless. The cypress runner would make it obvious that the test failed because of an attribute check on line x of file y, and then you go to that line and change the attribute. 

That said, if you decided to use both, then you still might want to leave some attribute checks in Cypress. It doesn't really add any extra time to the Cypress run. The expensive part of Cypress is loading the page and dynamic changes, not asserting the html.

Reply:
Maybe the feedback speed that a person needs relates to some function of their character & their attention span? I can tell you that I start getting annoyed/start yelling at my computer at about 2 seconds in. 😅 Or I get distracted. 🦋 🌈 

At any rate, 7 seconds per component is far too slow for my personal dev process. There are currently 38 components. at 7 seconds per test, we would be looking at over 4 minutes to run these tests, not including Elm compile time/other build steps.

As a comparison, if Elm code is already compiled, running _all_ of the monolith Elm tests is done in seconds:
<img width="682" alt="elm test is fast if elm code is already compiled" src="https://user-images.githubusercontent.com/8811312/163287341-cd833b55-936f-4f55-8c67-34b1483d6ce9.png">

I 100% agree that we need some UI testing. Percy & visual diffing in particular is awesome! But I think that if Deque's axe tooling became available for the Elm ecosystem, I would switch over to using Elm axe tests pretty much instantly in order to get the fast feedback loop. It makes _such_ a difference to me when doing TDD to have a near-instantaneous correctness check!

Reply:
Where is the time spent in that 4 seconds? Is it loading a browser? Could we keep it open?

Reply:
@tesk9 I see your point. I used to follow a similar test-driven development cycle at a previous job, and it was a lot of fun. I stopped doing that when I switched to writing Selenium code because it's absurdly slow. I miss being able to run the entire test suite instantly. 

Ok, I think I'm being overbearing about replacing unit tests. I would say it's a "both, and" situation. 

I think my goal would be to write a bunch of cypress tests that are easy to maintain (by removing the duplicate checks that you've indicated, among other things). That solves the problem we have in QA with needing UI test coverage on all the components. It also provides a tool that we can use in this repo as a final check, if not a development tool. I also firmly believe that the best way to orchestrate the percy snapshots axe checks is to do it component by component, rather than in one big batch. 

The proposition (reloaded): I give you a bunch of easy-to-maintain tests, and we get peace of mind on QA. How's that sound? 

-- 

Not-directly-relevant sidebar about speed and developer ergonomics. 

To answer your question @BrianHicks, I would say it's a combination of initializing the browser and loading the page. You can't skip the init browser step unless you jammed your tests into one giant `it` block. And that would still leave the page loading problem, which is multiplied by every `visit` command. You can save session data (cookies, etc) and preload application data (with a `before` hook), but it will always reload the browser at the start of each test. Even if we could do it, I don't think we should. Without isolation, UI tests are a lot less reliable, and a lot harder to maintain. The longer my test is, the longer I have to wait before it throws the error that I'm trying to fix (violation of fail-fast). 

The better way to improve test execution is via parallelization. This is offered via the Cypress Dashboard, but you can also [setup your own using Docker](https://stackoverflow.com/questions/61973532/cypress-parallelisation-without-cypress-dashboard). In theory [we've already done this on QA](https://github.com/NoRedInk/cypress/#install-and-set-up-docker-optional), but we never use it because we don't need to run the full suite during development. We just run the single test that we're working on. Then, if we need to, we'll run the entire suite via the Cypress Dashboard before merging a pull request. Also on every QA cycle, of course. 

With parallelism, you could probably lower that 4 minutes that Tessa mentioned down to a minute. 240 seconds / 8 parallel threads = 30 seconds altogether. Double that for safety and you've got one minute. 

Moreover, I would say that you don't actually need to run the full suite if you're only working on a single component. Even if you change a shared module, you can still just keep running the one test until you're completely finished, then run the full suite to ensure that everything else is still working.

All that said, I think that Tessa's point is still true. Even with a perfect setup, the developer experience would still be degraded without access to a suite a unit tests. 

So, let's do the "both, and" approach instead.

Reply:
Doing both elm tests and browser tests sounds good to me!

Separately, I've folded the axe checks into the percy test suite in https://github.com/NoRedInk/noredink-ui/pull/907, which means that now, every component is actually tested with axe (instead of just having the index page checked). Some rules are skipped, but I'll fix them separately.  (This is the work I mentioned [here in slack](https://noredink.slack.com/archives/C02NVG4M45U/p1649782033475539?thread_ts=1649720419.942899&cid=C02NVG4M45U)).

>  You can't skip the init browser step unless you jammed your tests into one giant it block.

This is how the tests work currently. It is indeed a giant it block! And it is not super zippy 😅 

> The longer my test is, the longer I have to wait before it throws the error that I'm trying to fix (violation of fail-fast).

I'm optimistic that we can improve the dev experience here by threading through some command line args -- like a filter by name for which components to run the test for.

--

I'm not opposed to Cypress by any means but I do have some requirements that would need to be met before I'd want to switch over:
1. I don't want to need to add a file or do _anything_ when I add a new component example. I want automatic snapshots and automatic axe checks.
2. I need to be able to run the test suite locally without the percy snapshots actually being taken. Percy is slow and I don't want me or other devs to worry about the PERCY_TOKEN or PERCY_BRANCH locally
3. I need to get actionable feedback from axe failures
4. I need fast feedback loops
5. I need to trust that `shake` will do it all -- as a dev, I shouldn't need to learn one-off script names right away and as gh actions, I need everything to just work

Reply:
Confused. Your list of needs is for which of the following scenarios?
1. Cypress only
    - (replacing all other testing)
2. Cypress & Elm
    - (replacing puppeteer only, including axe and percy)
3. Cypress & Elm & Puppeteer
    - (Cypress for UI functional testing only, not axe or percy) 

I'm leaning towards option 2, and I think I can give you everything that you need for that to work (including your requirements 1, 2, 3, and 5).

Reply:
List of needs is for any testing scenario! I don't feel at all strongly about what tooling we use -- I care most about the items I mentioned. 

FWIW on the performance front, currently seeing these results from Puppeteer for all the tests when skipping the Percy snapshots and running `time ./script/puppeteer-tests-no-percy.sh`
```
./script/puppeteer-tests-no-percy.sh  1.86s user 0.40s system 17% cpu 13.285 total
```

Running just the tests for `Switch`:
```
./script/puppeteer-tests-no-percy.sh Switch  0.47s user 0.10s system 25% cpu 2.262 total
```

So not super fast by any means, but the local dev experience isn't painful and it's feasible to run everything locally.